File: application/controllers/api/v1/bot/overdose.py
# Copyright (C) 2021-2025 tiksan
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

import json

from flask import request
from peewee import DoesNotExist
from tornium_commons.models import Faction, Server, ServerOverdoseConfig

from controllers.api.v1.bot.config import jsonified_server_config
from controllers.api.v1.decorators import ratelimit, session_required
from controllers.api.v1.utils import api_ratelimit_response, make_exception_response


@session_required
@ratelimit
def set_overdose_channel(guild_id: int, faction_id: int, *args, **kwargs):
    data = json.loads(request.get_data().decode("utf-8"))
    key = f"tornium:ratelimit:{kwargs['user'].tid}"

    channel_id = data.get("channel")

    if channel_id == "0" or channel_id == 0:
        channel_id = None
    if channel_id is not None and (channel_id in ("", 0) or not channel_id.isdigit()):
        return make_exception_response("1002", key)

    try:
        guild: Server = Server.select().where(Server.sid == guild_id).get()
    except DoesNotExist:
        return make_exception_response("1001", key)

    if kwargs["user"].tid not in guild.admins:
        return make_exception_response("4020", key)
    elif faction_id not in guild.factions:
        return make_exception_response("4021", key)

    try:
        faction: Faction = Faction.select(Faction.guild).where(Faction.tid == faction_id).get()
    except DoesNotExist:
        return make_exception_response("1102", key)

    if guild.sid != faction.guild_id:
        return make_exception_response("4021", key)

    ServerOverdoseConfig.create_or_update(guild_id, faction_id, channel=channel_id)

    return jsonified_server_config(guild), 200, api_ratelimit_response(key)


@session_required
@ratelimit
def set_overdose_policy(guild_id: int, faction_id: int, *args, **kwargs):
    data = json.loads(request.get_data().decode("utf-8"))
    key = f"tornium:ratelimit:{kwargs['user'].tid}"

    policy = data.get("policy")

    if policy not in ("immediate", "daily"):
        return make_exception_response("1000", key)

    try:
        guild: Server = Server.select().where(Server.sid == guild_id).get()
    except DoesNotExist:
        return make_exception_response("1001", key)

    if kwargs["user"].tid not in guild.admins:
        return make_exception_response("4020", key)
    elif faction_id not in guild.factions:
        return make_exception_response("4021", key)

    try:
        faction: Faction = Faction.select(Faction.guild).where(Faction.tid == faction_id).get()
    except DoesNotExist:
        return make_exception_response("1102", key)

    if guild.sid != faction.guild_id:
        return make_exception_response("4021", key)

    ServerOverdoseConfig.create_or_update(guild_id, faction_id, policy=policy)

    return jsonified_server_config(guild), 200, api_ratelimit_response(key)

File: application/static/faction/overdose.js
/* Copyright (C) 2021-2025 tiksan

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>. */

window.viewerErrorCallback = function (jsonError, container) {
    const errorNode = document.createElement("span");
    errorNode.textContent = " Data failed to load...";

    const errorIcon = document.createElement("i");
    errorIcon.classList.add("fa-solid", "fa-skull-crossbones");

    container.innerHTML = "";
    container.appendChild(errorIcon);
    container.appendChild(errorNode);
};

window.addOverdoseEventToViewer = function (event, eventContainer) {
    const eventNode = document.createElement("div");
    eventNode.classList.add("card", "mx-2", "mt-2", "viewer-card");
    eventNode.setAttribute("data-event-id", event.id);
    eventContainer.append(eventNode);

    const eventRow = document.createElement("div");
    eventRow.classList.add("row", "p-2", "align-middle");
    eventNode.append(eventRow);

    const eventNameElement = document.createElement("div");
    eventNameElement.classList.add("col-sm-12", "col-md-4");
    eventNameElement.textContent = `${event.user.name} [${event.user.id}]`;
    eventRow.append(eventNameElement);

    const eventActionElement = document.createElement("div");
    eventActionElement.classList.add("col-sm-12", "col-md-2");
    eventActionElement.textContent = event.action;
    eventRow.append(eventActionElement);

    const eventItemElement = document.createElement("div");
    eventItemElement.classList.add("col-sm-12", "col-md-3", "text-truncate");
    eventItemElement.textContent = event.drug == null ? "Unknown Drug" : event.drug.name;
    eventRow.append(eventItemElement);

    const eventDateTimeElement = document.createElement("time");
    eventDateTimeElement.classList.add("col-sm-12", "col-md-2");
    eventDateTimeElement.setAttribute("datetime", new Date(event.timestamp).toISOString());
    eventDateTimeElement.textContent = reltime(event.timestamp);
    eventRow.append(eventDateTimeElement);
};

window.filterTomSelectMultiple = function (element) {
    return Array.from(element.selectedOptions).map(({ value }) => value);
};

ready(() => {
    document.querySelectorAll(".automatic-tom-select").forEach((element) => {
        new TomSelect(element, {
            create: false,
            maxOptions: null,
            onItemAdd: tomselectClearAfterAdd,
        });
    });
    itemsRequest().finally(() => {
        document.querySelectorAll(".item-selector").forEach((element) => {
            new TomSelect(element, {
                create: false,
                onItemAdd: tomselectClearAfterAdd,
            });
        });
    });
});
